docs/Technical-Documentation/tech-memos/database-duplicate-detection.md

```diff
@@ -0,0 +1,344 @@
+- **Reduced complexity**: Eliminates complex sorting and removal logic that was required for in-memory duplicate detection
+
+```python
+# Current approach
+class SortedRecords:
+    ...
+
+# New simplified approach
+class Records:
+    """Maintains a dict of records where Key=Model and Value=[Record]."""
+
+    def __init__(self):
+        self.cases = dict()
+
+    def add_record(self, case_id, record_model_pair, line_num):
+        """Add a record_doc_pair to the dict."""
+        record, model = record_model_pair
+        if case_id is not None:
+            self.cases.setdefault(model, dict())[record] = None
+        else:
+            logger.error(f"Error: Case id for record at line #{line_num} was None!")
+
+    def get_bulk_create_struct(self):
+        """Return dict of form {document: {record: None}} for bulk_create_records to consume."""
+        return self.cases
+
+    def clear(self, all_created):
+        """Reset the dict if all records were created."""
+        if all_created:
+            # We don't want to re-assign self.cases here because we lose the keys of the record/doc types we've already
+            # made. If we don't maintain that state we might not delete everything if we need to roll the records back
+            # at the end of, or during parsing.
+            for key in self.cases.keys():
+                self.cases[key] = {}
+```
+
+#### Memory and Performance Benefits
+- **Reduced memory footprint**: Elimination of sorted collections significantly reduces memory usage during parsing
+- **Simplified data flow**: Records flow directly from parsing to database without complex in-memory manipulations
+- **Improved garbage collection**: Fewer complex data structures reduce pressure on Python's garbage collector
+
+### Database-Based Duplicate Detection Architecture
+
+The proposed implementation leverages PostgreSQL's efficient querying capabilities to detect duplicates after records have been persisted to the database, rather than maintaining in-memory caches during parsing.
+
+#### Exact Duplicate Detection
+
+The `_delete_exact_dups()` method below implements a slightly naive exact duplicate detection using database queries. This function does not take into account the queryset size brought into memory for processing which could cause problems in the event the entire datafile was duplicates.
+
+```python
+def _delete_exact_dups(self):
+    """Delete exact duplicate records from the DB."""
+    for schemas in self.schema_manager.schema_map.values():
+        schema = schemas[0]
+
+        fields = [f.name for f in schema.fields if f.name != "BLANK"]
+        duplicate_vals = (
+            schema.model.objects.filter(datafile=self.datafile)
+            .values(*fields)
+            .annotate(row_count=Count("line_number", distinct=True))
+            .filter(row_count__gt=1)
+        )
+
+        for dup_vals_dict in duplicate_vals:
```

Comment:
having a bit of trouble fully understanding this query - do we essentially create a `values_list` for every serialized record, then for each check if there's duplicates (and, presumably, do it all again for partial duplicates)? not sure if there's an efficiency concern (and i can't think of a better way off the top of my head)